src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,320 @@
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private unsafe static ulong CurlReceiveBodyCallback(
+            byte* buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                // Set task completion after all headers have been received
+                if (!state.Task.IsCompleted && !state.Task.IsCanceled && !state.Task.IsFaulted)
+                {
+                    state.TrySetResult(state.ResponseMessage);
+                }
+
+                // TODO: Handle compression and chunked-mode
+
+                // Wait for a reader
+                CurlResponseStream contentStream = state.ResponseMessage.ContentStream;
+                contentStream.WaitAndSignalReaders(buffer, (long)size);
+                return size;
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+                return (size > 0) ? size - 1 : 1; // Force curl_easy_perform to fail
+            }
+        }
+
+        private static ulong CurlSendCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+                    Stream contentStream = state.RequestContentStream;
+
+                    // TODO: Handle chunked-mode
+
+                    byte[] byteBuffer = state.RequestContentBuffer;
+                    int numBytes = contentStream.Read(state.RequestContentBuffer, 0, Math.Min(byteBuffer.Length, (int)totalSize));
+                    if (numBytes > 0)
+                    {
+                        Marshal.Copy(byteBuffer, 0, ptr, numBytes);
+                    }
+                    return (ulong)numBytes;
+                });
+        }
+
+        private static int CurlSendIoCtlCallback(
+            IntPtr handle,
+            int cmd,
+            IntPtr context)
+        {
+            if (cmd != curliocmd.CURLIOCMD_RESTARTREAD)
+            {
+                return curlioerr.CURLIOE_UNKNOWNCMD;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                if (state.CancellationToken.IsCancellationRequested)
+                {
+                    HandleAsyncException(state, null);
+                    return curlioerr.CURLIOE_FAILRESTART; // Force curl_easy_perform to fail
```

Comment:
Is the comment still accurate?  We're not longer using curl_easy_perform, right?  I see the same comment in a few other places.